llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -22142,6 +22142,10 @@ static SDValue performVectorShiftCombine(SDNode *N,
     if (DCI.DAG.ComputeNumSignBits(Op.getOperand(0)) > ShiftImm)
       return Op.getOperand(0);
 
+  // If the right shift is exact, the shifted out bits matter.
+  if (N->getOpcode() == AArch64ISD::VASHR && N->getFlags().hasExact())
```

Comment:
I think I have the same comments as David, I can see that CreateLShr also has the flag isExact.
It will not be triggered by your patch in : https://github.com/llvm/llvm-project/pull/92528, because you only add 
auto *AShr = Builder.CreateAShr(Cast, 24, "", true);
But, maybe this patch should also test for that one.